File: temalabor.py
import requests
import json
import time
from datetime import datetime, timedelta

budapest_timezone = timedelta(hours=1)
repo_adatok_file = 'repo_adatok.json'

def read_repo_adatok():
    try:
        with open(repo_adatok_file, 'r') as file:
            repo_adatok = json.load(file)
        return repo_adatok
    except FileNotFoundError:
        print(f'A(z) {repo_adatok_file} fájl nem található.')
        return None
    except json.JSONDecodeError:
        print(f'A(z) {repo_adatok_file} fájl nem érvényes JSON formátumot tartalmaz.')
        return None

def get_commit_details(api_url, github_token, owner, repo, sha):
    commit_url = f'{api_url}/repos/{owner}/{repo}/commits/{sha}'
    headers = {'Authorization': f'token {github_token}'}
    response = requests.get(commit_url, headers=headers)
    return response.json()

def main():
    repo_adatok = read_repo_adatok()

    if not repo_adatok:
        return

    repo_owner = repo_adatok.get('repo_owner')
    repo_name = repo_adatok.get('repo_name')
    github_token = repo_adatok.get('github_token')

    if not (repo_owner and repo_name and github_token):
        print('Hiányzó adatok a repo_adatok.json fájlban.')
        return

    api_url = 'https://api.github.com'

    while True:
        response = requests.get(f'{api_url}/repos/{repo_owner}/{repo_name}/commits', headers={'Authorization': f'token {github_token}'})
        commits = response.json()

        if len(commits) > 0:
            simplified_commits = []
            contributor_commits = {}

            for commit in commits:
                contributor_name = commit['commit']['author']['name']  # Fix this line
                commit_sha = commit['sha']

                # Részletes commit információk lekérése
                commit_details = get_commit_details(api_url, github_token, repo_owner, repo_name, commit_sha)

                # Felhasználjuk az author és committer mezőket a változott sorok számának meghatározására
                lines_changed = commit_details['stats']['total'] if commit_details.get('stats') else 0

                simplified_commit = {
                    'committer': contributor_name,
                    'lines_changed': lines_changed,
                    'commit message': commit['commit']['message'],
                    'timestamp': convert_utc_to_budapest(commit['commit']['author']['date'])  # Fix this line
                }
                simplified_commits.append(simplified_commit)

                # Update contributor_commits dictionary
                if contributor_name in contributor_commits:
                    contributor_commits[contributor_name] += 1
                else:
                    contributor_commits[contributor_name] = 1

            with open('changes.json', 'w') as json_file:
                json.dump(simplified_commits, json_file, indent=2)
            with open('contributors.json', 'w') as contributors_file:
                json.dump(contributor_commits, contributors_file, indent=2)

            print('JSON fájlok frissítve.')

        time.sleep(5)

def convert_utc_to_budapest(utc_timestamp):
    utc_time = datetime.strptime(utc_timestamp, '%Y-%m-%dT%H:%M:%SZ')
    budapest_time = utc_time + budapest_timezone
    return budapest_time.strftime('%Y-%m-%d %H:%M:%S')

if __name__ == "__main__":
    main()

File: temalabor.ts
import * as fs from 'fs';
import * as path from 'path';
import * as request from 'request-promise-native'; // Make sure to install the 'request-promise-native' package

const budapestTimezone: number = 1; // Adjust according to your timezone
const repoAdatokFile: string = 'repo_adatok.json';

function readRepoAdatok(): any | null {
    const filePath = path.join(__dirname, repoAdatokFile);

    try {
        const fileContent = fs.readFileSync(filePath, 'utf8');
        const repoAdatok = JSON.parse(fileContent);
        return repoAdatok;
    } catch (error) {
        if (error.code === 'ENOENT') {
            console.log(`${repoAdatokFile} file not found.`);
        } else {
            console.error(`Error reading ${repoAdatokFile} file: ${error.message}`);
        }
        return null;
    }
}

async function getCommitDetails(apiUrl: string, githubToken: string, owner: string, repo: string, sha: string): Promise<any> {
    const commitUrl = `${apiUrl}/repos/${owner}/${repo}/commits/${sha}`;
    const headers = {
        'Authorization': `token ${githubToken}`,
        'User-Agent': 'GitHubExtentionProject' // Replace with an appropriate User-Agent
    };

    try {
        const response = await request.get(commitUrl, { headers, json: true });
        return response;
    } catch (error) {
        console.error(`Error fetching commit details: ${error.message}`);
        return null;
    }
}

function convertUtcToBudapest(utcTimestamp: string): string {
    const utcTime = new Date(utcTimestamp);
    const budapestTime = new Date(utcTime.getTime() + budapestTimezone * 60 * 60 * 1000);
    return budapestTime.toISOString().replace(/T/, ' ').replace(/\..+/, '');
}

export async function jsonGenerator() {
    const repoAdatok = readRepoAdatok();

    if (!repoAdatok) {
        return;
    }

    const repoOwner = repoAdatok.repo_owner;
    const repoName = repoAdatok.repo_name;
    const githubToken = repoAdatok.github_token;

    if (!repoOwner || !repoName || !githubToken) {
        console.log('Missing data in repo_adatok.json file.');
        return;
    }

    const apiUrl: string = 'https://api.github.com';

    while (true) {
        try {
            const commits = await request.get(`${apiUrl}/repos/${repoOwner}/${repoName}/commits`, {
                headers: {
                    'Authorization': `token ${githubToken}`,
                    'User-Agent': 'GitHubExtentionProject'
                },
                json: true
            });

            if (commits.length > 0) {
                const simplifiedCommits = [];
                const contributorCommits = {};

                for (const commit of commits) {
                    const contributorName = commit.commit.author.name; // Fix this line
                    const commitSha = commit.sha;

                    const commitDetails = await getCommitDetails(apiUrl, githubToken, repoOwner, repoName, commitSha);

                    const linesChanged = commitDetails.stats.total || 0;

                    const simplifiedCommit = {
                        committer: contributorName,
                        lines_changed: linesChanged,
                        commit_message: commit.commit.message,
                        timestamp: convertUtcToBudapest(commit.commit.author.date) // Fix this line
                    };

                    simplifiedCommits.push(simplifiedCommit);

                    if (contributorName in contributorCommits) {
                        contributorCommits[contributorName] += 1;
                    } else {
                        contributorCommits[contributorName] = 1;
                    }
                }

                fs.writeFileSync('changes.json', JSON.stringify(simplifiedCommits, null, 2));
                fs.writeFileSync('contributors.json', JSON.stringify(contributorCommits, null, 2));

                console.log('JSON files updated.');
            }

            await new Promise(resolve => setTimeout(resolve, 5000)); // Sleep for 5 seconds
        } catch (error) {
            console.error(`Error in main loop: ${error.message}`);
        }
    }
}


if (require.main === module) {
    jsonGenerator();
}
